Explain event callbacks in "Monitoring for parameter changes" (#5543)
ros2/ros2_documentation: source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst

            self.callback_handle = self.handler.add_parameter_callback(
                parameter_name="an_int_param",

>                 node_name="node_with_parameters",

Reviewer: > In the original pasted Python code, specifying node_name="node_with_parameters" to self.handler.add_parameter_callback is redundant.

can you elaborate a bit this? i do not see why this is redundant...

https://github.com/ros2/rclpy/blob/3414456ddaf1163d1951d301336028c9600bd58e/rclpy/rclpy/parameter_event_handler.py#L292

it looks like required argument here to specify this test node's parameter event activity.

Author: Oh true, thanks for catching this! I was looking at the rclcpp version, where it is optional.

https://github.com/ros2/rclcpp/blob/4b1c125/rclcpp/include/rclcpp/parameter_event_handler.hpp#L235-L252

I'll revert, but maybe as a separate issue it'd be nice to make the API identical between rclpy and rclcpp.

The callback we set previously in the node has been invoked and has displayed the new updated value.
You can now terminate the running parameter_event_handler sample using ^C in the terminal.

Extensions
----------

So far, we built and tested a small node that monitors a single parameter owned by the node itself.
Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.

Monitor changes to another node's parameters

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: i think original doc explains well, especially we can find the following description. i would keep here as it is, since i do not really see the difference.

```rst
You can also use the ParameterEventHandler to monitor parameter changes to another node's parameters.
Let's update the SampleNodeWithParameters class to monitor for changes to a parameter in another node.
```

Author: Thanks for the review! Just to check my understanding, you're saying to revert this block of additions (Your text in code block appears just below these lines), right? 

I realize I didn't make the intent very clear (what does ["lift 'remote node'  one level higher"](https://github.com/ros2/ros2_documentation/pull/5543/commits/5c57c8ae67c7e7a38361c3c4e709831bde101e6c) mean?). This aims to make the "Tasks" section only talk about the basic usage of `add_parameter_callback`, and delegate further discussion to a separate "Extensions" section (at the same level as "Tasks").

So we'd go from:

> * Background
> * Prerequisites
> * Tasks
>   - 1 Create a package
>   - 2 Write the Python node
>   - 3 Build and run **<-- subsections (3.1, 3.2) are hidden from the ToC**
> * Summary
> * Related content

To:

> * Background
> * Prerequisites
> * Tasks
>   - 1 Create a package
>   - 2 Write the Python node
>   - 3 Build and run <-- no more subsections, it only talks about the basic usage
> * **Extensions**
>   - **Monitor changes to another node's parameters**
>   - **Monitor all node parameters simultaneously**
> * Summary
> * Related content

If this was already understood and you still prefer to revert I'm fine, just making sure. Personally I think having these visible in the ToC is nicer.

It's also a bit surprising for "_Monitor changes to another node's parameters_" and "_Monitor all node parameters simultaneously_" to be subsections of something called "_Build and run_"

Reviewer: I see, thanks for your explanation. i think your proposed sections are better, i will go ahead to resolve the related comments.

The callback we set previously in the node has been invoked and has displayed the new updated value.
You can now terminate the running parameter_event_handler sample using ^C in the terminal.

> 

Reviewer 2: Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

Extensions
----------

So far, we built and tested a small node that monitors a single parameter owned by the node itself.

> Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.

Reviewer 2: ```suggestion
Using this node as a base, two other usecases where the ParameterEventHandler can be useful are presented below.
```

> 

Reviewer 2: Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

      [INFO] [1699821958.757770223] [node_with_parameters]: Received an update to parameter: a_double_param: 3.45

> Monitor all node parameters simultaneously

Reviewer: this is great enhancement 🚀 to explain user how to use `add_parameter_event_callback`.
can we sectionize this?

```suggestion
3.2 Monitor all node parameters simultaneously
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Author: See the previous comment.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> If you have a node with lots of parameters, it would be cumbersome to have to call ``add_parameter_callback`` once for each parameter.

Reviewer: ```suggestion
If you need to monitor multiple nodes or parameters at the same time, it would be cumbersome to have to call ``add_parameter_callback`` once for each of them.
```

Author: I might be wrong, but it looks to me that `add_parameter_event_callback` can only be used to monitor self-owned parameters (in Python at least), not remote ones. Can you confirm?

Reviewer: i do not think that limitation, it just fetches the event data from `/parameter_events` topic, and issue the callback. (rclcpp has the consistent behavior with rclpy, as far as i see.)

Author: Ah yes, I should have looked at the implementation! Thanks for the suggestion.

      [INFO] [1699821958.757770223] [node_with_parameters]: Received an update to parameter: a_double_param: 3.45

> 

Reviewer 2: Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

Monitor all node parameters simultaneously
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you need to monitor multiple nodes or parameters at the same time, it would be cumbersome to have to call ``add_parameter_callback`` once for each of them.

> In this case, you can use ``add_parameter_event_callback`` to register a single callback that fires when *any* of the node's parameters change.

Reviewer 2: Does `add_parameter_event_callback()` only trigger the given callback when a specific node's parameters change, or does it trigger it when _any_ parameters of _any_ nodes change?

Author: The latter. I pushed 9d63c3840bc8ecbaa9e95b97b74fad31738dad17 to make this more explicit.

To do this, first update the SampleNodeWithParameters constructor to add the following code:

.. code-block:: Python

    def __init__(...):
        self.declare_parameter("another_double_param", 0.0)
        ...
        self.event_calback_handle = self.handler.add_parameter_event_callback(
            callback=self.event_callback,
        )

This declares a new double parameter ``another_double_param`` and adds an event callback that will monitor both parameters.
The event callback signature is different from that of regular single-parameter callbacks, so we need to define a suitable callback as well:

.. code-block:: Python

    def event_callback(self, parameter_event):
        self.get_logger().info("Received parameter event")

        for p in parameter_event.changed_parameters:
            self.get_logger().info(
                f"Inside event: {p.name} changed to: {rclpy.parameter.parameter_value_to_python(p.value)}"
            )

> 

Reviewer 2: ```suggestion
```

> Note that the ``parameter_event`` is of type `rcl_interfaces/ParameterEvent <https://docs.ros.org/en/rolling/p/rcl_interfaces/msg/ParameterEvent.html>`_.

Reviewer 2: ```suggestion
Note that the ``parameter_event`` is of type {interface(rcl_interfaces/msg/ParameterEvent)}.
```

Although it's not shown in this tutorial, event callbacks can also be used to monitor when parameters are added or deleted.

> 

Reviewer 2: ```suggestion
```

Navigate back to the root of your workspace, ``ros2_ws``, and rebuild your updated package as before:

.. code-block:: console

    $ colcon build --packages-select python_parameter_event_handler

Then source the setup files:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

      $ . install/setup.bash

  .. group-tab:: macOS

    .. code-block:: console

      $ . install/setup.bash

  .. group-tab:: Windows

    .. code-block:: console

      $ call install\setup.bat

> 

Reviewer 2: ```suggestion
```

To test the new event callback, first run the parameter_event_handler node:

.. code-block:: console

     $ ros2 run python_parameter_event_handler node_with_parameters

> Then, from a second terminal (with ROS initialized), let's set the original int parameter:

Reviewer 2: I'm assuming you mean "sourcing," so it would be better to say that, since that's what we call it everywhere:

```suggestion
Then, from a second terminal (with ROS sourced), let's set the original int parameter:
```

.. code-block:: console

     $ ros2 param set node_with_parameters an_int_param 44

Upon executing this command, you should see both the single-parameter callback, as well as the event callback being fired:

.. code-block:: console

      [INFO] [1746414766.240101027] [node_with_parameters]: Received an update to parameter: an_int_param: 44
      [INFO] [1746414766.243499816] [node_with_parameters]: Received parameter event
      [INFO] [1746414766.244271445] [node_with_parameters]: Inside event: an_int_param changed to: 4

> Now set the new doble parameter:

Author: Oops. Should we add `doble->double` to https://github.com/ros2/ros2_documentation/blob/rolling/codespell_dictionary.txt ? :smile: 
```suggestion
Now set the new double parameter:
```

.. code-block:: console

     $ ros2 param set node_with_parameters another_double_param 4.4

Since no single-parameter callback was added (via ``add_parameter_callback``) for the double parameter, we should see only the event callback fire:

.. code-block:: console

      [INFO] [1746414962.604832196] [node_with_parameters]: Received parameter event
      [INFO] [1746414962.607429035] [node_with_parameters]: Inside event: another_double_param changed to: 4.4

.. note::

   When setting multiple parameters at once, it's best to use ``set_parameters_atomically``, explained in :doc:`../../Concepts/Basic/About-Parameters`.
   This way, the event callback is only fired once.

> 

Reviewer 2: ```suggestion
```
